blogposts/2021/how-we-built-batch-changes.md

```diff
@@ -0,0 +1,213 @@
+built the product, you don't _really_ know what it should look like or how to
+best build it. You've never used it, so how would you know how it best works?
+You also haven't built it (yet), so you don't have experience building it.
+Chances that you can accurately predict how its development will go are low.
+
+Acknowledging that you _don't_ plan the development of the product from start to
+finish in every detail. You _don't_ nail down and define the hundreds of steps
+it will take to go from 0 to fully-built and shipped and launched product.
+Instead, working iteration-based, you define roughly where you want to go and
+then take a single small step. Then you look around, get your bearings, correct
+course if necessary, and take another step. Build something, ship it, learn,
+build again.
+
+I personally love working like this and think that in a lot of contexts in which
+software is being developed (fast-moving startups, for example) it's the best way to
+ship successful products.
+
+But there's one huge problem with this approach: it's really easy to iterate
+yourself into a corner. If all you do is take small steps you can miss the point
+where you should take big leaps to someplace else entirely. You end up with [a
+local maximum](https://elezea.com/2012/09/iphone-5-local-maximum/). On top of
+that: even if you realise that you're stuck and _know_ that you must take a big
+leap, it's hard to abandon the things you've built in many, many small
+iterations. "So much work went into this! We can't simply delete it! Right?"
+Right, it's the [sunken cost
+fallacy](https://en.wikipedia.org/wiki/Sunk_cost#Fallacy_effect).
+
+Back to the five pull requests and how we've managed to defy both of these dangers.
+
+In September 2019 we started working on Batch Changes, except that back then it
+was called _Automation_ and it was already implemented, kinda: [sqs](TODO) had
+built an impressive prototype to show to customers and ask them whether they'd
+use something like this. More than one said "yes, I would, and _exactly like
+that_". "[Build one to throw away](https://wiki.c2.com/?PlanToThrowOneAway)" is
+what they say - but do you? Do you really throw away a prototype that's more
+than 10k lines of code and that customers pointed at and said "I want this"?
+
+We did, with [this pull request][pr1], which is a placeholder for all of the
+code we wrote in September and October 2019 to rebuild the functionality of the
+prototype from scratch. Why did we do it? We thought that building the product
+from the ground up, fully understanding and owning the code is more important to
+the long-term success of the project and the team than merging a prototype to
+get something out there as fast as possible.
+
+Only a few months later, though, we were stuck. We realised that what we had
+built and shipped as an alpha to customers just didn't work on a conceptual
+level. In that version, campaigns (now called batch changes) were executed on
```

Comment:
The next question that could be answered here is `What went "wrong" (though it's not really wrong) in the product development that you built something that didn't work on a conceptual level?`

Reply:
Ah, yes. That _is_ interesting. I'm not sure I have an answer though (or if someone else could better write that), but I'll think about it.

Reply:
I can't up with something that I can confidently write about. I'll keep thinking about it though.